### DISCUSSION TITLE
`Tabs`: Implement developer-friendly api

### DISCUSSION THREAD

**Discussion by with-heart** [chakra-ui/chakra-ui#4162] [Ideas]
After some discussion in our core team meeting yesterday, we decided we'd like to change the `Tabs` component api. 

## Problems

- painful to coordinate indexes between `Tab`+`TabPanel` combinations when tabs are dynamic
- can't customize component tree because props are passed directly rather than shared via context (see #3045)

## Solution

We'd like to restructure the implementation and api of the `Tabs` component to resolve those problems.

- `Tabs` is indexed by `string` ids
	- `Tab` relates to `TabPanel` via `panelId` prop
	- `TabPanel` relates to `Tabs` via `id` prop
- `Tabs` data is shared via descendant context instead of passed props

```tsx
const App = () => {
  return (
    <Tabs defaultTab="a">
		<TabList>
			{/* can be nested! */}
			<div>
				<Tab panelId="a">A</Tab>
			</div>
			<Tab panelId="b">B</Tab>
			<Tab panelId="c">C</Tab>
		</TabList>
		<TabPanels>
			{/* can be nested! */}
			<div>
				<TabPanel id="a">A panel!</TabPanel>
			</div>
			<TabPanel id="b">B panel!</TabPanel>
			<TabPanel id="c">C panel!</TabPanel>
		</TabPanels>
	</Tabs>
  )
}
```

**Comment by aboveyunhai** [comment]
@with-heart  glad that you bring it up.

I also want to ask if it's possible to bring the original "wrong" implementation of `onChange` back as a option. `onChange` only fires when index changes. 
I tired to use `onClick` to replace `onChange` due to some functionalities, but `onClick` does not contain the index by default, and I found it's hard to get the current active index via `doucment.getElementby...()`, 
Also, the fundamental part: `onChange` listen to `tabIndex` but `onClick` is limited by clicking behavior.
```
onTabChange( (index) => {
  if(index === tabIndex) return;
})
//this is equivalent to current onChange();
```

or make `function onChange(index, default=true) // when set to false means it will be fired even click the same tab`

So it allows user to retrig some custom functions even when click the same tabs. 
It can be extremely handy when there are multiple tabs with it's own dynamic content, and you only want to refetch data when click the same tabs again as a shortcut behavior.

**Comment by martinezleoml** [comment]
@with-heart It would be very useful to wrap a disabled `Tab` into a `Tooltip` so we can give end-users explanations for example. 🙂

**Comment by with-heart** [comment]
@aboveyunhai I think I understand what you're asking, but would like to make sure. Could you create a CodeSandbox example of what that looks like?

**Comment by aboveyunhai** [comment]
> @aboveyunhai I think I understand what you're asking, but would like to make sure. Could you create a CodeSandbox example of what that looks like?

I just create a simple minimal example with a "not working" example but pretty much summarizing the idea;

https://codesandbox.io/s/magical-cori-idbhf?file=/src/index.tsx

The behavior can be more interesting for example if user goes to the page by url like `samplepage?tab=1` then `setTabIndex` and then trig `fetch` for the tab, the ability to control both index from "onChange" (this actually goes against the name) can do more flexible stuffs like controlling the `fetch/refetch` behaviors depending on the libs.
`onClick` is quite limit due to its own focus.

**Comment by kodell** [comment]
This would be a huge help, dynamic tabs are a mess especially when trying to map to routes. That is:
1. I want to set `defaultIndex` on mount based on the requested route.
2. I want my URL route to update when I switch tabs.

A new `defaultPanel` prop would take care of (1).

Updating the signature of `onChange` to be `(index: number, panelId: string | undefined) => void` would take care of (2) and remain backwards compatible.

**Comment by vitordino** [comment]
pretty nice suggestion!
looks a bit similar to [reakit’s `<Tab/>` api](https://reakit.io/docs/tab/), but they use a custom hook instead of context.

**Comment by segunadebayo** [comment]
Permit me to convert this to a discussion since it's not technically an issue.

I love that API as well and I've been thinking about it some more. This API extends to the `Accordion` component as well, basically, we don't want to use `index` in the userland 'coz it seems like an implementation detail leak.

As an analogy, the platform uses `id` to associate elements, think of `<label htmlFor />` + `<input id />`  and also `<button form="#id">` + `<form id="" />`. Maybe we can leverage these thoughts and patterns as well in v2.

**Comment by aboveyunhai** [comment]
@with-heart  just a out of scope question regarding `Tabs` and possible other UI components.
Does Chakra UI team ever consider to add "simple" gesture support, for example swipe to change tabs. But I guess it might cause a lot of unexpected issues and larger the bundle size potentially.
This is probably more related to Chakra UI future roadmap. It's really hard to define the meaning of UI.
The reason I asked that is because I recently just add the simple feature into the tabs for touch device.

Or maybe just leave a example spot for this kind of implementation, and keep the simplicity of the library itself.

**Comment by jdaly13** [comment]
Was this every implemented - I am struggling with dynamic tabs with the current implementation - looking at the docs there's no examples of dynamic tabs - for example when I create a new tab and set tabIndex it still goes to old one
